features/storage/kvstore/tdbstore/TDBStore.cpp

```diff
@@ -0,0 +1,1378 @@
+            ih->new_key = false;
+        } else if (ret == MBED_ERROR_ITEM_NOT_FOUND) {
+            if (create_flags & delete_flag) {
+                goto fail;
+            }
+            if (_num_keys >= _max_keys) {
+                increment_max_keys();
+            }
+            ih->new_key = true;
+        } else {
+            goto fail;
+        }
+        ih->bd_base_offset = _free_space_offset;
+
+        check_erase_before_write(_active_area, ih->bd_base_offset, rec_size);
+    }
+
+    ret = MBED_SUCCESS;
+
+    // Fill handle and header fields
+    // Jump to offset after header (header will be written at finalize phase)
+    ih->bd_curr_offset = ih->bd_base_offset + align_up(sizeof(record_header_t), _prog_size);
+    ih->offset_in_data = 0;
+    ih->hash = hash;
+    ih->ram_table_ind = ram_table_ind;
+    ih->header.magic = tdbstore_magic;
+    ih->header.header_size = sizeof(record_header_t);
+    ih->header.revision = tdbstore_revision;
+    ih->header.flags = create_flags;
+    ih->header.key_size = strlen(key);
+    ih->header.reserved = 0;
+    ih->header.data_size = final_data_size;
+    // Calculate CRC on header and key
+    ih->header.crc = crc32(initial_crc, sizeof(record_header_t) - sizeof(ih->header.crc), &ih->header);
+    ih->header.crc = crc32(ih->header.crc, ih->header.key_size, key);
+
+    // Write key now
+    os_ret = write_area(_active_area, ih->bd_curr_offset, ih->header.key_size, key);
+    if (os_ret) {
+        ret = MBED_ERROR_WRITE_FAILED;
+        goto fail;
+    }
+    ih->bd_curr_offset += ih->header.key_size;
+    goto end;
+
+fail:
+    // mark handle as invalid by clearing magic field in header
+    ih->header.magic = 0;
+    _mutex.unlock();
+
+end:
+    return ret;
+}
+
+int TDBStore::set_add_data(set_handle_t handle, const void *value_data, size_t data_size)
+{
+    int os_ret, ret = MBED_SUCCESS;
+    inc_set_handle_t *ih;
+
+    if (handle != _inc_set_handle) {
+        return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!value_data && data_size) {
+        return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    ih = reinterpret_cast<inc_set_handle_t *> (handle);
+
+    if (!ih->header.magic) {
+        return MBED_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (ih->offset_in_data + data_size > ih->header.data_size) {
+        ret = MBED_ERROR_INVALID_SIZE;
+        // Need GC as otherwise our storage is kept in a non-usable state
+        garbage_collection();
+        goto fail;
+    }
+
+    // Update CRC with data chunk
+    ih->header.crc = crc32(ih->header.crc, data_size, value_data);
+
+    // Write data chunk
+    os_ret = write_area(_active_area, ih->bd_curr_offset, data_size, value_data);
+    if (os_ret) {
+        ret = MBED_ERROR_WRITE_FAILED;
+        goto fail;
+    }
+    ih->bd_curr_offset += data_size;
+    ih->offset_in_data += data_size;
+    goto end;
+
+fail:
+    // mark handle as invalid by clearing magic field in header
+    ih->header.magic = 0;
+    _mutex.unlock();
```

Comment:
Do we want any function other than finalize to unlock the mutex?
a. We said that when current call will overflow data size , user may still call set_add_data again or set_finalize.
b. Even on fatal fail user might call finalize and cause problem